mne/decoding/tests/test_ssd.py

```diff
@@ -0,0 +1,240 @@
+
+    Spectral ratio measure for best n_components selection
+    See :footcite:`NikulinEtAl2011`, Eq. (24).
+    """ 
+    psd, freqs = psd_array_welch(
+    X_ssd, sfreq=sf, n_fft=sf)
+    sig_idx = _time_mask(freqs, *ssd.freqs_signal)
+    noise_idx = _time_mask(freqs, *ssd.freqs_noise)
+    if psd.ndim == 3:
+        mean_sig = psd[:, :, sig_idx].mean(axis=2).mean(axis=0)
+        mean_noise = psd[:, :, noise_idx].mean(axis=2).mean(axis=0)
+        spec_ratio = mean_sig / mean_noise
+    else:
+        mean_sig = psd[:, sig_idx].mean(axis=1)
+        mean_noise = psd[:, noise_idx].mean(axis=1)
+        spec_ratio = mean_sig / mean_noise
+    sorter_spec = spec_ratio.argsort()[::-1]
+    return spec_ratio, sorter_spec
+
+def simulate_data(freqs_sig = [9, 12], n_trials=100, n_channels=20,
+                  n_samples=500, samples_per_second=250,
+                  n_components=5, SNR=0.05, random_state=42):
+    """Simulate data according to an instantaneous mixin model.
+
+    Data are simulated in the statistical source space, where n=n_components
+    sources contain the peak of interest,
+    """
+    rs = np.random.RandomState(random_state)
+
+    filt_params_signal=dict(l_freq=freqs_sig[0], h_freq=freqs_sig[1],
+                                  l_trans_bandwidth=1, h_trans_bandwidth=1,
+                                  fir_design='firwin')
+
+    # generate a orthogonal mixin matrix
+    mixing_mat = np.linalg.svd(rs.randn(n_channels, n_channels))[0]
+    # define sources
+    S_s = rs.randn(n_trials*n_samples, n_components)
+    # filter source in the specific freq. band of interest
+    S_s = filter_data(S_s.T, samples_per_second, **filt_params_signal ).T
+    S_n = rs.randn(n_trials*n_samples, n_channels-n_components)
+    
+    S=np.hstack((S_s, S_n))
+    # mix data
+    X_s = np.dot(mixing_mat[:,:n_components], S_s.T).T   
+    X_n = np.dot(mixing_mat[:,n_components:], S_n.T).T 
+    # add noise
+    X_s = X_s / np.linalg.norm(X_s, 'fro')
+    X_n = X_n / np.linalg.norm(X_n, 'fro')
+    X = SNR * X_s + (1-SNR) * X_n
+    X=X.T
+    S=S.T
+    return X, mixing_mat, S
+
+@pytest.mark.slowtest
+def test_ssd():
+    """Test Common Spatial Patterns algorithm on raw data."""
+    X, A, S = simulate_data()
+    sf=250
+    n_channels = X.shape[0]
+    info=create_info(ch_names=n_channels, sfreq=sf, ch_types='eeg')
+    n_components_true=5
+    
+    # Init
+    # data type
+    filt_params_signal = dict(l_freq=freqs_sig[0], h_freq=freqs_sig[1],
+                                  l_trans_bandwidth=1, h_trans_bandwidth=1,
+                                  fir_design='firwin')
+    filt_params_noise = dict(l_freq=freqs_noise[0], h_freq=freqs_noise[1],
+                                 l_trans_bandwidth=1, h_trans_bandwidth=1,
+                                 fir_design='firwin')
+    ssd = SSD(info, filt_params_signal, filt_params_noise)
+    # components no int
+    pytest.raises(ValueError, ssd, n_components='foo')
```

Comment:
I think this should be SSD, not ssd. The instantiated object cannot be called. This test is trying to instantiate the test. I'd suggest something like:

```pytest.raises(ValueError, SSD, filt_params_signal, filt_params_noise, n_components='foo')```

Reply:
Just a drive by comment to mention that this is the old way to write these, context managers are preferred as they're nicer more explicit:
```
with pytest.raises(Value error, match='must be an int'):
    SSD(...)
```
Also the error here should be TypeError. You'll get this for free if you use _ensure_int. Also make use of _check_option and _validate_type where possible. I can give more specific inline guidance if this is too vague

Reply:
@vpeterson more homework by @larsoner :)

Reply:
thank you both !. But, anyway, this is weird, this is not the last version I pushed 10 days ago. I will try to push it again. I don't understand what had happened.